### ISSUE TITLE
Considering a safer null handling

### ISSUE THREAD

**Issue by sroddy** [google/built_value.dart#306]
I have seen that the Quiver library contains an Optional implementation (https://docs.flutter.io/flutter/quiver.core/Optional-class.html) that would be amazing to use together with built_value.

It would be much less error-prone to handle nullable variables this way instead of the current way, as you would be aware in any part of the code of the optional nature of the value itself.

How I imagine it is the following:

```dart
class BuiltClass extends BuiltValue<BuiltClass, BuiltClassBuilder> {
  ...
  Optional<String> get anOptionalString;
}

...
final valueWithOptional = new BuiltClassBuilder()..build();
assert(!valueWithOptional.anOptionalString.isPresent);
final anotherValue = valueWithOptional.rebuild((b) => b..anOptionalString = "a string");
assert(anotherValue.anOptionalString.isPresent);
assert(anotherValue.anOptionalString.value == "a string");
```

What do you think about this? Do you think it is feasible?

**Comment by zoechi** [comment]
Non-nullable types are still on the roadmap for Dart (at some point after 2.0.0), this should fix that.

**Comment by dave26199** [comment]
Funny story: I added Optional to Quiver :) ... for a use very like what you
suggest. But...

The reason it's not supported is I wanted to wait for Non Nullable By
Default (NNBD). It will be easier to migrate from nullable to NNBD than
from Optional to NNBD.

But, NNBD is taking longer than I'd hoped. It may be worth reconsidering at
some point. Not sure yet. A third possibility that I'd like to explore
first is whether we can add built_value-specific checks for nullable in the
style of NNBD via an analyzer plugin. I.e. warn about unsafe use.

Tl;dr: yes, definitely aiming for something better here. It's just not
clear what. Input welcome :)

Günter Zöchbauer <notifications@github.com> schrieb am Mi., 27. Dez. 2017,
19:05:
